### PR TITLE
ADAPT-3096 Create Text Field component

### PR DESCRIPTION
**Pull request by rebeccahongsf**
# READY FOR REVIEW

# Summary
- Create Text Field component

# Needed By (Date)
- tbd

# Urgency
- medium-high

# Steps to Test

1. Check out branch locally
2. View local preview or [Netlify](https://deploy-preview-1--decanter-react-forms.netlify.app/?path=/story/composite-text-field--default) and confirm the following design tidbits (see [Figma mockup](https://www.figma.com/file/Kmd4utmJFPRMVeCFEEBQhLtx/Decanter-Design-System?node-id=10694%3A292))
- Text should be ~18px
- Input padding should be 19px left, top, right and 20px bottom.
- Icon should have two position options — Left or Right
- Error label, input, and message should be `digital-red-light`
- Valid label, input, and message should be `digital-green`
- Custom Icons should work (see [Custom Input with Icon Preview](https://deploy-preview-1--decanter-react-forms.netlify.app/?path=/story/composite-text-field--user-input))
And check to see if anything else appears out of place. 😄 

**Additional Notes:** 
- Disable state is currently just background color `black-10` since the diagonal stripes cannot be implemented through tailwind utility class styles.
- Currently the Char count is live update. I have not tested it for c&p reactions nor determined how we would handle errors in the event that characters exceed the limit/limit has been reached.
- [Password Error and Validation checklist](https://www.figma.com/file/Jk1dfcFHSA59ljplAUEo2C/My-Account-Wireframes-2021?node-id=1037%3A11547) has not been implemented. I plan to initiate further discussion on how we would like to handle that (either through an existing library (e.g. [react-password-checklist](https://www.npmjs.com/package/react-password-checklist)) or creating a password specific component) during ADAPT Dev Chat 

4. Review the code!
I anticipate that some conditional statements might need to be refactored. Please do not hesitate to tag me, ask any questions, and suggest additional examples/stories to be added. Thank you!!! 🥳 

# Associated Issues and/or People
- [ADAPT-3096](https://stanfordits.atlassian.net/browse/ADAPT-3096)

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

**Comment by yvonnetangsu** [comment]
@rebeccahongsf These looks great. We talked about form elements in dev chat yesterday, and looks like we have some decisions to make. @mattanglin thinks that the form components and the styling can go into Decanter React main, while DERF imports them and hook them up with React Hook Forms. @sherakama thinks that we have some experimentation and research to do before we decide. However, 2 decisions came out of the meeting:

- We'll be adding JIT mode to Decanter React and tell people that they'll need a compiler
- We'll not care about the out of the box HTML form elements and just assume that people will add su- classnames if they want the proper styling
😄 

@sherakama @mattanglin Please add if I miss anything. Thanks!

**Comment by rebeccahongsf** [comment]
Thank you @yvonnetangsu! This makes so much sense 😄 Bringing these components into Decanter-React sounds like a good plan. I'll open PRs for the form components (TextField and ErrorBox) in Decanter-React. I'll also leave these two PRs open just in case we decide to go in a different direction.

I've created this ticket to do more research about how we would create a Decanter React Forms library incorporated with react-hook-forms: [ADAPT-3249](https://stanfordits.atlassian.net/browse/ADAPT-3249)

**Comment by yvonnetangsu** [comment]
> Thank you @yvonnetangsu! This makes so much sense 😄 Bringing these components into Decanter-React sounds like a good plan. I'll open PRs for the form components (TextField and ErrorBox) in Decanter-React. I'll also leave these two PRs open just in case we decide to go in a different direction.
> 
> I've created this ticket to do more research about how we would create a Decanter React Forms library incorporated with react-hook-forms: [ADAPT-3249](https://stanfordits.atlassian.net/browse/ADAPT-3249)

Yeah, your work is not lost! 😄  And there is a chance we'll keep these here. I think I'm going to write some Tickets for our FFW friends to help explore this first.

They look great btw! 👏🏼

**Comment by sherakama** [comment]
This looks amazing at first pass. Thanks for this work.
One small comment, the date type should probably be out on its own component as that one is different enough from a text input field that conceptually it makes sense to separate. I also suspect we will get into different options for it at some point.

**Comment by sherakama** [comment]
This is on hold until we decide on the approach for RHF integration.

File: src/TextField/TextField.js [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    validText,
+    isError,
+    isValid,
+    isIcon,
+    iconPosition,
+    isDisabled,
+    isRequired,
+    showCharCount,
+  },
+  props
```
Comment (mattanglin) [inline src/TextField/TextField.js:33]: Is this correct to have additional props coming in as a second parameter to the component function? I've not come across this pattern before.

File: src/TextField/TextField.js [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+            iconInputStyle
+          )}
+          type={type}
+          placeholder={placeholder}
+          maxLength={maxLength ?? null}
+          minLength={minLength ?? null}
+          disabled={isDisabled}
+          required={isRequired}
+          onChange={handleOnChange}
+          {...props}
```
Comment (mattanglin) [inline src/TextField/TextField.js:120]: Assuming that the consumer of this component passes their own `onChange` handler wouldn't the above `handleOnChange` get overwritten?

**Pull request closed without merging**